pkg/cache/cache.go

```diff
@@ -0,0 +1,92 @@
+package cache
+
+import (
+	"fmt"
+	"os/user"
+	"sync"
+	"time"
+)
+
+var DefaultCacheExpiration = time.Now().Add(30 * time.Minute)
+
+//go:generate mockgen -source=cache.go -package=cache -destination=./cache_mock.go
+type RosaCache interface {
+	Set(k string, x interface{}, d time.Time)
+	Get(k string) (interface{}, bool)
+	Items() map[string]Item
+	Dir() (string, error)
+}
+
+var _ RosaCache = &rosaCache{}
+
+type RosaCacheSpec struct {
+	DefaultExpiration time.Time
+}
+
+type rosaCache struct {
+	defaultExpiration time.Time
+	items             map[string]Item
+	mu                sync.RWMutex
+}
+
+func NewRosaCache(spec RosaCacheSpec) RosaCache {
+	if !spec.DefaultExpiration.IsZero() {
+		DefaultCacheExpiration = spec.DefaultExpiration
+	}
+	return &rosaCache{
+		defaultExpiration: DefaultCacheExpiration,
+		items:             make(map[string]Item),
+	}
+}
+
+func (c *rosaCache) Set(k string, x interface{}, d time.Time) {
+	if d.IsZero() {
+		d = c.defaultExpiration
+	}
+	c.mu.Lock()
+	c.items[k] = Item{
+		Object:     x,
+		Expiration: d,
+	}
+	c.mu.Unlock()
+}
+
+func (c *rosaCache) Get(k string) (interface{}, bool) {
+	c.mu.RLock()
+	item, found := c.items[k]
+	if !found {
+		c.mu.RUnlock()
+		return nil, false
+	}
+	if !item.Expiration.IsZero() {
+		if item.Expired() {
+			c.mu.RUnlock()
+			return nil, false
+		}
+	}
+	c.mu.RUnlock()
+	return item.Object, true
+}
+
+func (c *rosaCache) Items() map[string]Item {
+	c.mu.RLock()
+	defer c.mu.RUnlock()
+	m := make(map[string]Item, len(c.items))
+	for k, v := range c.items {
+		if !v.Expiration.IsZero() {
+			if v.Expired() {
+				continue
+			}
+		}
+		m[k] = v
+	}
+	return m
+}
+
+func (c *rosaCache) Dir() (string, error) {
```

Comment:
@ciaranRoche I'd be tempted for us to introduce support for an ENV variable here that allows for the cache directory to be overridden, something like `ROSA_CONFIG_DIR`. So the flow would be:

- Check `ROSA_CONFIG_DIR` exists, if yes, use that; Otherwise
- Use `~/.config/ocm/cache.gob`; Otherwise
- Use `~/tmp/ocm/cache.gob`

That type of variable tends to be useful in CI runs where the home directory of a container is ephemeral between runs, but we can bind mount some non-ephemeral storage onto the `ROSA_CONFIG_DIR` location. It's also just useful for folks who don't want this stuff in their home directory.

If none of those work, then caching is disabled. I think otherwise we run the risk of leaving `cache.gob` files in directories where the command was run from, which I'm not sure feels like being a good citizen.

WDYT?